histogram/src/sliding_window.rs

```diff
@@ -0,0 +1,470 @@
+        self.tick_to(instant);
+
+        let index = self.config.value_to_index(value)?;
+
+        self.live[index].fetch_add(count, Ordering::Relaxed);
+
+        Ok(())
+    }
+
+    /// Returns a snapshot that covers the provided range. Both the start and
+    /// end of the range will be adjusted to the proceeding snapshot (tick)
+    /// boundary. This results in distribution between the start and end times,
+    /// which includes the provided start but excludes the provided end.
+    pub fn snapshot_between(
+        &self,
+        range: core::ops::Range<UnixInstant>,
+    ) -> Result<crate::Snapshot, Error> {
+        self.snapshot();
+
+        let tick_at = self.tick_at();
+
+        if range.start < self.started {
+            return Err(Error::OutOfSlidingWindow);
+        }
+
+        // convert unix times to monotonic clock times
+        let start = self.tick_origin + (range.start - self.started - self.interval);
+        let end = self.tick_origin + (range.end - self.started - self.interval);
+
+        // lookup snapshot information
+        let start = self.snapshot_info(start, tick_at)?;
+        let end = self.snapshot_info(end, tick_at)?;
+
+        let mut total_count = 0_u128;
+
+        let buckets: Vec<u64> = self.snapshots[start.index]
+            .iter()
+            .zip(self.snapshots[end.index].iter())
+            .map(|(start, end)| {
+                let count = end
+                    .load(Ordering::Relaxed)
+                    .wrapping_sub(start.load(Ordering::Relaxed));
+                total_count += count as u128;
+                count
+            })
+            .collect();
+
+        let histogram = crate::Histogram {
+            config: self.config,
+            total_count,
+            buckets: buckets.into(),
+        };
+
+        Ok(Snapshot {
+            range: start.range.start..end.range.end,
+            histogram,
+        })
+    }
+
+    /// Returns the current inclusive range of time covered by the histogram.
+    pub fn range(&self) -> Range<UnixInstant> {
+        let elapsed = self.tick_at.load(Ordering::Relaxed) - self.interval - self.tick_origin;
+        let end = self.started + elapsed;
+        let start = end - self.span;
+
+        start..end
+    }
+
+    /// Moves the window forward, if necessary.
+    fn tick_to(&self, instant: Instant) {
+        loop {
+            let tick_at = self.tick_at.load(Ordering::Relaxed);
+
+            // fast path when the window does not need to be advanced
+            if instant < tick_at {
+                return;
+            }
+
+            // otherwise we need to slide the window forward
+
+            // To actually snapshot, let's just move the tick_at forward to
+            // unblock other increments. This will slightly smear things into
+            // the snapshot that occur after the end boundary, but this
+            // trade-off seems worth it to reduce pause duration.
```

Comment:
I'm not sure if this is the right behavior in all cases. It works for the common case of frequently updated (and queried) histograms.

However, if we're behind by multiple ticks (but less than the buffer length) we will have two or more threads snapshotting in parallel.

If we're behind by more than the ring buffer size, this is might be really broken.

Reply:
And also fixed by simplifying.